flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java

```diff
@@ -0,0 +1,130 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.highavailability;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.jobmaster.JobResult;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/** An abstract class for threadsafe implementations of the {@link JobResultStore}. */
+public abstract class AbstractThreadsafeJobResultStore implements JobResultStore {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(AbstractThreadsafeJobResultStore.class);
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    @Override
+    public void createDirtyResult(JobResultEntry jobResultEntry) throws IOException {
+        Preconditions.checkState(
+                !hasJobResultEntry(jobResultEntry.getJobId()),
+                "Job result store already contains an entry for for job %s",
+                jobResultEntry.getJobId());
+
+        readWriteLock.writeLock().lock();
+        try {
+            createDirtyResultInternal(jobResultEntry);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract void createDirtyResultInternal(JobResultEntry jobResultEntry)
+            throws IOException;
+
+    @Override
+    public void markResultAsClean(JobID jobId) throws IOException, NoSuchElementException {
+        if (hasCleanJobResultEntry(jobId)) {
+            LOG.debug("The job {} is already marked as clean. No action required.", jobId);
+            return;
+        }
+
+        readWriteLock.writeLock().lock();
+        try {
+            markResultAsCleanInternal(jobId);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract void markResultAsCleanInternal(JobID jobId)
+            throws IOException, NoSuchElementException;
+
+    @Override
+    public boolean hasJobResultEntry(JobID jobId) throws IOException {
+        readWriteLock.readLock().lock();
+        try {
+            return hasDirtyJobResultEntry(jobId) || hasCleanJobResultEntry(jobId);
```

Comment:
```suggestion
            return hasDirtyJobResultEntryInternal(jobId) || hasCleanJobResultEntryInternal(jobId);
```
nit: not really necessary because of the lock being reentrant. But using the `*Internal` methods would avoid going through the lock routine once more here. On our case it's not really a performance issue because the method won't be called that regularly. I just wanted to point it out...